amm-executor/src/Spectrum/Executor/EventSource/Stream.hs

```diff
@@ -0,0 +1,182 @@
+  )
+import Spectrum.Executor.EventSource.Persistence.LedgerHistory
+  ( LedgerHistory (..), mkLedgerHistory )
+import Spectrum.Executor.EventSource.Data.TxEvent
+  ( TxEvent(AppliedTx, UnappliedTx) )
+import Spectrum.Executor.EventSource.Data.TxContext
+  ( TxCtx(LedgerTx) )
+import Spectrum.LedgerSync.Data.LedgerUpdate
+  ( LedgerUpdate(RollForward, RollBackward) )
+import Ouroboros.Consensus.Block
+  ( Point )
+import Spectrum.Executor.EventSource.Persistence.Data.BlockLinks
+  ( BlockLinks(BlockLinks, txIds, prevPoint) )
+import Spectrum.Executor.EventSource.Persistence.Config
+  ( LedgerStoreConfig )
+
+newtype EventSource s m = EventSource
+  { upstream :: s m (TxEvent 'LedgerTx)
+  }
+
+mkEventSource
+  :: forall m s env.
+    ( IsStream s
+    , Monad (s m)
+    , MonadAsync m
+    , MonadResource m
+    , MonadReader env m
+    , HasType (MakeLogging m m) env
+    , HasType EventSourceConfig env
+    , HasType LedgerStoreConfig env
+    )
+  => LedgerSync m
+  -> m (EventSource s m)
+mkEventSource lsync = do
+  mklog@MakeLogging{..}      <- askContext
+  EventSourceConfig{startAt} <- askContext
+  lhcong                     <- askContext
+
+  logging     <- forComponent "EventSource"
+  persistence <- mkLedgerHistory mklog lhcong
+
+  seekToBeginning logging persistence lsync startAt
+  pure $ EventSource $ upstream' logging persistence lsync
+
+upstream'
+  :: forall s m. (IsStream s, Monad (s m), MonadAsync m)
+  => Logging m
+  -> LedgerHistory m
+  -> LedgerSync m
+  -> s m (TxEvent 'LedgerTx)
+upstream' logging@Logging{..} persistence LedgerSync{..}
+  = S.repeatM pull >>= processUpdate logging persistence
+  & S.trace (infoM . show)
+
+processUpdate
+  :: forall s m.
+    ( IsStream s
+    , Monad (s m)
+    , MonadIO m
+    , MonadBaseControl IO m
+    , MonadThrow m
+    )
+  => Logging m
+  -> LedgerHistory m
+  -> LedgerUpdate Block
+  -> s m (TxEvent 'LedgerTx)
+processUpdate
+  _
+  LedgerHistory{..}
+  (RollForward (BlockAlonzo (ShelleyBlock (Ledger.Block (TPraos.BHeader hBody _) txs) hHash))) =
+    let
+      txs'  = txSeqTxns txs
+      point = ConcretePoint (TPraos.bheaderSlotNo hBody) (ConcreteHash ch)
+        where ch = OneEraHash . toShort . CC.hashToBytes . TPraos.unHashHeader . unShelleyHash $ hHash
+    in S.before (setTip point)
+      $ S.fromFoldable txs' & S.map (AppliedTx . fromAlonzoLedgerTx hHash)
+processUpdate logging lh (RollBackward point) = streamUnappliedTxs logging lh point
+processUpdate Logging{..} _ upd = S.before (errorM $ "Cannot process update " <> show upd) mempty
+
+streamUnappliedTxs
+  :: forall s m.
+    ( IsStream s
+    , Monad (s m)
+    , MonadIO m
+    , MonadBaseControl IO m
+    , MonadThrow m
+    )
+  => Logging m
+  -> LedgerHistory m
+  -> Point Block
+  -> s m (TxEvent 'LedgerTx)
+streamUnappliedTxs Logging{..} LedgerHistory{..} point = join $ S.fromEffect $ do
+  knownPoint <- pointExists $ fromPoint point
+  let
+    rollbackOne :: ConcretePoint -> s m (TxEvent 'LedgerTx)
```

Comment:
What's going on if (un)applied transactions will be lost (in case of handler error for example)? We setup new point right here before any conformation from handlers. 
Or we don't care about it?